cudax/samples/vector_add_async/async_uninitialized_buffer.cuh

```diff
@@ -0,0 +1,208 @@
+      : __mr_(__mr)
+      , __count_(__count)
+      , __buf_(__count == 0
+                 ? nullptr
+                 : static_cast<_Tp*>(__mr.allocate_async(__count * sizeof(_Tp), alignof(max_align_t), __stream)))
+      , __alloc_stream_(__stream)
+  {}
+
+  async_uninitialized_buffer(
+    _Tp, _CUDA_VMR::async_resource_ref<_Properties...> __mr, const size_t __count, ::cuda::stream_ref __stream)
+      : async_uninitialized_buffer(__mr, __count, __stream)
+  {}
+
+  async_uninitialized_buffer(const async_uninitialized_buffer&)            = delete;
+  async_uninitialized_buffer& operator=(const async_uninitialized_buffer&) = delete;
+
+  //! @brief Move construction
+  //! @param __other Another \c async_uninitialized_buffer
+  async_uninitialized_buffer(async_uninitialized_buffer&& __other) noexcept
+      : __mr_(__other.__mr_)
+      , __count_(_CUDA_VSTD::exchange(__other.__count_, {}))
+      , __buf_(_CUDA_VSTD::exchange(__other.__buf_, {}))
+      , __alloc_stream_(_CUDA_VSTD::exchange(__other.__alloc_stream_, {}))
+  {}
+
+  async_uninitialized_buffer& operator=(async_uninitialized_buffer&& __other) noexcept
+  {
+    if (this != _CUDA_VSTD::addressof(__other))
+    {
+      async_uninitialized_buffer(_CUDA_VSTD::move(__other)).swap(*this);
+    }
+    return *this;
+  }
+
+  ~async_uninitialized_buffer()
+  {
+    if (__buf_)
+    {
+      __mr_.deallocate_async(__buf_, __count_ * sizeof(_Tp), alignof(max_align_t), __alloc_stream_);
+    }
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE pointer begin() const noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns an aligned pointer to end of the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE pointer end() const noexcept
+  {
+    return __get_data() + __count_;
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE pointer data() const noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns the size of the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr size_type size() const noexcept
+  {
+    return __count_;
+  }
+
+  //! @rst
+  //! Returns the :ref:`async_resource_ref <libcudacxx-extended-api-memory-resources-resource-ref>` used to allocate
+  //! the buffer
+  //! @endrst
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE _CUDA_VMR::async_resource_ref<_Properties...> resource() const noexcept
+  {
+    return __mr_;
+  }
+
+  //! @brief Swaps the contents with those of another \c uninitialized_buffer
+  //! @param __other The other \c uninitialized_buffer.
+  _CCCL_HOST_DEVICE constexpr void swap(async_uninitialized_buffer& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__mr_, __other.__mr_);
+    _CUDA_VSTD::swap(__count_, __other.__count_);
+    _CUDA_VSTD::swap(__buf_, __other.__buf_);
+    _CUDA_VSTD::swap(__alloc_stream_, __other.__alloc_stream_);
+  }
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS // friend functions are currently brocken
```

Comment:
A minor spelling mistake in the comment:

```suggestion
#ifndef DOXYGEN_SHOULD_SKIP_THIS // friend functions are currently broken
```

Reply:
this is copied from elsewhere. i saw the misspelling and thought it was funny given the context so i left it. :-)